File: go/funding_raised_test.go
package fundingraised

import "testing"

func TestWhereGivenCompany(t *testing.T) {
	options := map[string]string{
		"company_name": "Facebook",
	}

	rows := len(Where(options))
	if rows != 7 {
		t.Error("Expected 7 rows. Got ", rows)
	}
}

func TestWhereCorrectKeys(t *testing.T) {
	options := map[string]string{
		"company_name": "Facebook",
	}

	rows := Where(options)
	row := rows[0]

	if row["permalink"] != "facebook" {
		t.Error("Expected permalink to be facebook. Got", row["permalink"])
	}

	if row["company_name"] != "Facebook" {
		t.Error("Expected company name to be Facebook. Got", row["company_name"])
	}

	if row["number_employees"] != "450" {
		t.Error("Expected number of employees to be 450. Got", row["number_employees"])
	}

	if row["category"] != "web" {
		t.Error("Expected category to be web. Got", row["category"])
	}

	if row["city"] != "Palo Alto" {
		t.Error("Expected city to be Palo Alto. Got", row["city"])
	}

	if row["state"] != "CA" {
		t.Error("Expected state to be CA. Got", row["state"])
	}

	if row["funded_date"] != "1-Sep-04" {
		t.Error("Expected funded date to be 1-Sep-04. Got", row["funded_date"])
	}

	if row["raised_amount"] != "500000" {
		t.Error("Expected raised amount to be 500000. Got", row["raised_amount"])
	}

	if row["round"] != "angel" {
		t.Error("Expected round to be angel. Got", row["round"])
	}
}

func TestWhereGivenCity(t *testing.T) {
	options := map[string]string{
		"city": "Tempe",
	}

	rows := len(Where(options))
	if rows != 3 {
		t.Error("Expected 3 rows having city == Tempe. Got", rows)
	}
}

func TestWhereGivenState(t *testing.T) {
	options := map[string]string{
		"state": "CA",
	}

	rows := len(Where(options))
	if rows != 873 {
		t.Error("Expected 873 rows having state == CA. Got", rows)
	}
}

func TestWhereGivenRound(t *testing.T) {
	options := map[string]string{
		"round": "a",
	}

	rows := len(Where(options))
	if rows != 582 {
		t.Error("Expected 582 rows having round == a. Got", rows)
	}
}

func TestWhereMultipleOptions(t *testing.T) {
	options := map[string]string{
		"company_name": "Facebook",
		"round":        "a",
	}

	rows := len(Where(options))
	if rows != 1 {
		t.Error("Expected 1 row having company name == Facebook and round == a. Got", rows)
	}
}

func TestWhereNonExistence(t *testing.T) {
	options := map[string]string{
		"company_name": "NotFacebook",
	}

	rows := len(Where(options))
	if rows != 0 {
		t.Error("Expected an empty list. Got", rows)
	}
}

func TestFindByGivenCompanyName(t *testing.T) {
	options := map[string]string{
		"company_name": "Facebook",
	}

	row, err := FindBy(options)

	if err != nil {
		t.Error("Expected FindBy to return no errors")
	}

	if row["permalink"] != "facebook" {
		t.Error("Expected permalink to be facebook. Got", row["permalink"])
	}

	if row["company_name"] != "Facebook" {
		t.Error("Expected company name to be Facebook. Got", row["company_name"])
	}

	if row["number_employees"] != "450" {
		t.Error("Expected number of employees to be 450. Got", row["number_employees"])
	}

	if row["category"] != "web" {
		t.Error("Expected category to be web. Got", row["category"])
	}

	if row["city"] != "Palo Alto" {
		t.Error("Expected city to be Palo Alto. Got", row["city"])
	}

	if row["state"] != "CA" {
		t.Error("Expected state to be CA. Got", row["state"])
	}

	if row["funded_date"] != "1-Sep-04" {
		t.Error("Expected funded date to be 1-Sep-04. Got", row["funded_date"])
	}

	if row["raised_amount"] != "500000" {
		t.Error("Expected raised amount to be 500000. Got", row["raised_amount"])
	}

	if row["raised_currency"] != "USD" {
		t.Error("Expected raised currency to be USD. Got", row["raised_currency"])
	}

	if row["round"] != "angel" {
		t.Error("Expected round to be angel. Got", row["round"])
	}
}

func TestFindByGivenState(t *testing.T) {
	options := map[string]string{
		"state": "CA",
	}

	row, err := FindBy(options)

	if err != nil {
		t.Error("Expected FindBy to return no errors")
	}

	if row["permalink"] != "digg" {
		t.Error("Expected permalink to be digg. Got", row["permalink"])
	}

	if row["company_name"] != "Digg" {
		t.Error("Expected company name to be Digg. Got", row["company_name"])
	}

	if row["number_employees"] != "60" {
		t.Error("Expected number of employees to be 60. Got", row["number_employees"])
	}

	if row["category"] != "web" {
		t.Error("Expected category to be web. Got", row["category"])
	}

	if row["city"] != "San Francisco" {
		t.Error("Expected city to be San Francisco. Got", row["city"])
	}

	if row["state"] != "CA" {
		t.Error("Expected state to be CA. Got", row["state"])
	}

	if row["funded_date"] != "1-Dec-06" {
		t.Error("Expected funded date to be 1-Dec-06. Got", row["funded_date"])
	}

	if row["raised_amount"] != "8500000" {
		t.Error("Expected raised amount to be 8500000. Got", row["raised_amount"])
	}

	if row["raised_currency"] != "USD" {
		t.Error("Expected raised currency to be USD. Got", row["raised_currency"])
	}

	if row["round"] != "b" {
		t.Error("Expected round to be b. Got", row["round"])
	}
}

func TestFindByMultipleOptions(t *testing.T) {
	options := map[string]string{
		"company_name": "Facebook",
		"round":        "c",
	}

	row, err := FindBy(options)

	if err != nil {
		t.Error("Expected FindBy to return no errors")
	}

	if row["permalink"] != "facebook" {
		t.Error("Expected permalink to be facebook. Got", row["permalink"])
	}

	if row["company_name"] != "Facebook" {
		t.Error("Expected company name to be Facebook. Got", row["company_name"])
	}

	if row["number_employees"] != "450" {
		t.Error("Expected number of employees to be 450. Got", row["number_employees"])
	}

	if row["category"] != "web" {
		t.Error("Expected category to be web. Got", row["category"])
	}

	if row["city"] != "Palo Alto" {
		t.Error("Expected city to be Palo Alto. Got", row["city"])
	}

	if row["state"] != "CA" {
		t.Error("Expected state to be CA. Got", row["state"])
	}

	if row["funded_date"] != "1-Oct-07" {
		t.Error("Expected funded date to be 1-Oct-07. Got", row["funded_date"])
	}

	if row["raised_amount"] != "300000000" {
		t.Error("Expected raised amount to be 300000000. Got", row["raised_amount"])
	}

	if row["raised_currency"] != "USD" {
		t.Error("Expected raised currency to be USD. Got", row["raised_currency"])
	}

	if row["round"] != "c" {
		t.Error("Expected round to be c. Got", row["round"])
	}
}

func TestFindByNonExistence(t *testing.T) {
	options := map[string]string{
		"company_name": "NotFacebook",
	}

	_, err := FindBy(options)

	if err == nil {
		t.Error("Expected FindBy to return an error")
	}
}

File: java/funding-raised/src/test/java/com/checkr/interviews/FundingRaisedTest.java
package com.checkr.interviews;
import java.util.*;
import java.io.IOException;

import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

/**
 * Unit test for simple App.
 */
public class FundingRaisedTest
    extends TestCase
{
    /**
     * Create the test case
     *
     * @param testName name of the test case
     */
    public FundingRaisedTest ( String testName )
    {
        super( testName );
    }

    /**
     * @return the suite of tests being tested
     */
    public static Test suite()
    {
        return new TestSuite( FundingRaisedTest.class );
    }

    /**
     * Rigourous Test :-)
     */
    public void testWhereGivenCompany() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("company_name", "Facebook");
            assertEquals(FundingRaised.where(options).size(), 7);
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testWhereGivenCity() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("city", "Tempe");
            assertEquals(FundingRaised.where(options).size(), 3);
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testWhereGivenState() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("state", "CA");
            assertEquals(FundingRaised.where(options).size(), 873);
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testWhereGivenRound() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("round", "a");
            assertEquals(FundingRaised.where(options).size(), 582);
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testMultipleOptions() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("round", "a");
            options.put("company_name", "Facebook");
            assertEquals(FundingRaised.where(options).size(), 1);
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testWhereNotExists() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("company_name", "NotFacebook");
            assertEquals(FundingRaised.where(options).size(), 0);
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testWhereCorrectKeys() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("company_name", "Facebook");
            Map<String, String> row = FundingRaised.where(options).get(0);

            assertEquals(row.get("permalink"), "facebook");
            assertEquals(row.get("company_name"), "Facebook");
            assertEquals(row.get("number_employees"), "450");
            assertEquals(row.get("category"), "web");
            assertEquals(row.get("city"), "Palo Alto");
            assertEquals(row.get("state"), "CA");
            assertEquals(row.get("funded_date"), "1-Sep-04");
            assertEquals(row.get("raised_amount"), "500000");
            assertEquals(row.get("round"), "angel");
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testFindByGivenCompanyName() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("company_name", "Facebook");
            Map<String, String> row = FundingRaised.findBy(options);

            assertEquals(row.get("permalink"), "facebook");
            assertEquals(row.get("company_name"), "Facebook");
            assertEquals(row.get("number_employees"), "450");
            assertEquals(row.get("category"), "web");
            assertEquals(row.get("city"), "Palo Alto");
            assertEquals(row.get("state"), "CA");
            assertEquals(row.get("funded_date"), "1-Sep-04");
            assertEquals(row.get("raised_amount"), "500000");
            assertEquals(row.get("round"), "angel");
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        } catch(NoSuchEntryException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testFindByGivenState() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("state", "CA");
            Map<String, String> row = FundingRaised.findBy(options);

            assertEquals(row.get("permalink"), "digg");
            assertEquals(row.get("company_name"), "Digg");
            assertEquals(row.get("number_employees"), "60");
            assertEquals(row.get("category"), "web");
            assertEquals(row.get("city"), "San Francisco");
            assertEquals(row.get("state"), "CA");
            assertEquals(row.get("funded_date"), "1-Dec-06");
            assertEquals(row.get("raised_amount"), "8500000");
            assertEquals(row.get("round"), "b");
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        } catch(NoSuchEntryException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testFindByMultipleOptions() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("company_name", "Facebook");
            options.put("round", "c");
            Map<String, String> row = FundingRaised.findBy(options);

            assertEquals(row.get("permalink"), "facebook");
            assertEquals(row.get("company_name"), "Facebook");
            assertEquals(row.get("number_employees"), "450");
            assertEquals(row.get("category"), "web");
            assertEquals(row.get("city"), "Palo Alto");
            assertEquals(row.get("state"), "CA");
            assertEquals(row.get("funded_date"), "1-Oct-07");
            assertEquals(row.get("raised_amount"), "300000000");
            assertEquals(row.get("round"), "c");
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        } catch(NoSuchEntryException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        }
    }

    public void testFindByNotExists() {
        try {
            Map<String, String> options = new HashMap<String, String> ();
            options.put("company_name", "NotFacebook");
            options.put("round", "c");
            Map<String, String> row = FundingRaised.findBy(options);
            fail("findBy should throw exception");
        } catch(IOException e) {
            System.out.print(e.getMessage());
            System.out.print("error");
        } catch(NoSuchEntryException e) {
        }
    }
}
